Userland/Libraries/LibAudio/WavLoader.cpp

```diff
@@ -6,53 +6,124 @@
  */
 
 #include "WavLoader.h"
-#include "Buffer.h"
+#include "LoaderError.h"
 #include <AK/Debug.h>
+#include <AK/Endian.h>
 #include <AK/FixedArray.h>
 #include <AK/NumericLimits.h>
-#include <AK/OwnPtr.h>
 #include <AK/Try.h>
-#include <LibCore/File.h>
-#include <LibCore/FileStream.h>
+#include <LibCore/MemoryStream.h>
 
 namespace Audio {
 
-static constexpr size_t maximum_wav_size = 1 * GiB; // FIXME: is there a more appropriate size limit?
+static constexpr size_t const maximum_wav_size = 1 * GiB; // FIXME: is there a more appropriate size limit?
 
 WavLoaderPlugin::WavLoaderPlugin(StringView path)
     : m_file(Core::File::construct(path))
 {
-    if (!m_file->open(Core::OpenMode::ReadOnly)) {
-        m_error = LoaderError { String::formatted("Can't open file: {}", m_file->error_string()) };
-        return;
-    }
-    m_stream = make<Core::InputFileStream>(*m_file);
 }
 
 MaybeLoaderError WavLoaderPlugin::initialize()
 {
-    if (m_error.has_value())
-        return m_error.release_value();
+    if (m_backing_memory.has_value())
+        m_stream = LOADER_TRY(Core::Stream::MemoryStream::construct(m_backing_memory.value()));
+    else
+        m_stream = LOADER_TRY(Core::Stream::File::open(m_file->filename(), Core::Stream::OpenMode::Read));
+
     TRY(parse_header());
     return {};
 }
 
 WavLoaderPlugin::WavLoaderPlugin(Bytes const& buffer)
+    : m_backing_memory(buffer)
+{
+}
+
+template<typename SampleReader>
+MaybeLoaderError WavLoaderPlugin::read_samples_from_stream(Core::Stream::Stream& stream, SampleReader read_sample, FixedArray<Sample>& samples) const
+{
+    switch (m_num_channels) {
+    case 1:
+        for (auto& sample : samples)
+            sample = Sample(LOADER_TRY(read_sample(stream)));
+        break;
+    case 2:
+        for (auto& sample : samples) {
+            auto left_channel_sample = LOADER_TRY(read_sample(stream));
+            auto right_channel_sample = LOADER_TRY(read_sample(stream));
+            sample = Sample(left_channel_sample, right_channel_sample);
+        }
+        break;
+    default:
+        VERIFY_NOT_REACHED();
+    }
+    return {};
+}
+
+// There's no i24 type + we need to do the endianness conversion manually anyways.
+static ErrorOr<double> read_sample_int24(Core::Stream::Stream& stream)
+{
+    u8 byte = 0;
+    TRY(stream.read(Bytes { &byte, 1 }));
+    u32 sample1 = byte;
+    TRY(stream.read(Bytes { &byte, 1 }));
+    u32 sample2 = byte;
+    TRY(stream.read(Bytes { &byte, 1 }));
+    u32 sample3 = byte;
+
+    i32 value = 0;
+    value = sample1 << 8;
+    value |= (sample2 << 16);
+    value |= (sample3 << 24);
+    return static_cast<double>(value) / static_cast<double>(NumericLimits<i32>::max());
```

Comment:
Why does this divide by the max of i32 instead of i24?

Reply:
That's a bug, I copied that from the LegacyBuffer implementation. Good catch!